dask/dataframe/io/parquet/arrow.py

```diff
@@ -749,17 +960,48 @@ def read_partition(
 
         dfs = []
         for rg in row_group:
-            piece = pq.ParquetDatasetPiece(
-                path,
-                row_group=rg,
-                partition_keys=partition_keys,
-                open_file_func=partial(fs.open, mode="rb"),
-            )
-            arrow_table = cls._parquet_piece_as_arrow(
-                piece, columns, partitions, **kwargs
-            )
+            if pa_ds is not None and isinstance(rg, pa_ds.ParquetFileFragment):
+                # `rg` is already a `ParquetFileFragment`, pyarrow
+                # knows how to convert this to a `table`
+                cols = []
+                for name in columns:
+                    if name is None:
+                        if "__index_level_0__" in schema.names:
+                            columns.append("__index_level_0__")
+                    else:
+                        cols.append(name)
+                arrow_table = rg.to_table(
+                    use_threads=False,
+                    schema=schema,
+                    columns=cols,
+                    # TODO: Modify tests to allow full filtering
+                    # filter=pq._filters_to_expression(filters) if filters else None,
+                )
+            else:
+                piece = pq.ParquetDatasetPiece(
+                    path,
+                    row_group=rg,
+                    partition_keys=partition_keys,
+                    open_file_func=partial(fs.open, mode="rb"),
+                )
+                arrow_table = cls._parquet_piece_as_arrow(
+                    piece, columns, partitions, **kwargs
+                )
             df = cls._arrow_table_to_pandas(arrow_table, categories, **kwargs)
 
+            # For pyarrow.dataset api, need to convert partition columns
+            # to categorigal manually...
```

Comment:
@jorisvandenbossche - When using the legacy api, we get a categrical dtype for "partitioned" columns.  Can we get the same format without the manual conversion used here?

Reply:
Yes, there is an option for this in the partitioning interface. See eg the PR where I added this to `pq.read_table`: https://github.com/apache/arrow/pull/7777, and see docs here: https://github.com/apache/arrow/blob/cb7d1c1e29d75ea91257a7b6ee0b652cd8b66a04/python/pyarrow/_dataset.pyx#L1335

With that option, you can enable to automatically dictionary encode (and thus become categorical on conversion to pandas) all partition fields that are strings. 
Other dtypes (typically ints) are not included, though. So if dask wants to keep returning categorical dtype for those as well, you will need to do that manually fo now (and if so, also welcome to open a JIRA about supporting ints). I think the reason we don't dictionary encode ints is because for those it doesn't really give any memory / efficiency increase (while for strings this is the case)